# SpongeMethylation

Files and code used for de Mendoza et al 2019 "Convergent evolution of a vertebrate-like methylome in a marine sponge". **Nature Ecology and Evolution** 2019 Oct;3(10):1464-1473. doi: 10.1038/s41559-019-0983-2

Free access to the manuscript: https://www.ncbi.nlm.nih.gov/pmc/articles/pmid/31558833/.

**Code**
* Master_Script_deMendozaNEE.R

**Files**
* `SequenceAlignments.zip`: Amino acid sequence alignments of DNMT1, DNMT3, TET and UHRF1.
* `Aqu2.1.genes.bed.gz`: *Amphimedon queenslandica* Aqu2.1 gene annotation.
* `Amphimedon_queenslandica.NNNNstrech.bed.gz`: Regions encompassing Ns (gaps) in the *Amphimedon* genome assembly.
* `bacterial_contigs`: list of *Amphimedon* assembly contigs from bacterial origin.
* `Amphimedon_UMR_LMR.zip`: Output files from MethylSeekR UMR/LMR finding for each developmental sample.
* `Amphimedon_DAPseq_TF_motifs.zip`: Best motif as identified by HOMER2 for *Amphimedon* Transcription Factors DAP-seq/ampDAP-seq experiments.
* `Amphimedon_DSS_DMRs.zip`: Output files from DSS DMR calling across all developmental stages.
* `CELSEQ_Aqu2.summarised.tsv.gz`: Mean CEL-seq TPM summarised values for the developmental stages, from GSE70185.
* `hmC_150windows_both_stages_01.bed.gz`: Regions with hmCG/CG >= 0.1 in both Precompetent larva and juvenile stages.